### PR TITLE
[SAFETY] redemption rate within safety thresholds

### PR DESCRIPTION
**Pull request by riley-stride**
## What is the purpose of the change

When liquid staking, an attacker would desire a low RR, as it would mint them more stAssets than deserved. 
When redeeming, an attacker would desire a high RR, as it would redeem them more native assets than deserved.

Here, we add two stakeibc params:
- `safety_min_redemption_rate_threshold`: we cut off liquid staking if the RR is below this threshold
- `safety_max_redemption_rate_threshold`: we cut off redeeming if the RR is above this threshold

These are parameters. That means we can change them with time. I set them to default, but `safety_max_redemption_rate_threshold` will have to change after ~1 year. 

For now I'm setting `safety_max_redemption_rate_threshold` to `3` (the only cosmos chain that might exceed this in 12 months is evmos). We will have to submit a gov proposal maybe once per year to increase this on a few chains. 

After launch, we should maybe move these parameters to live on host zones themselves, so that we can set them specific to each zone (e.g. ATOM's would be far LT evmos').

## Brief Changelog

- add both params to `stakeibc/params.go` and `params.proto`
- add two helpers to `stakeibc/keeper/keeper.go` to check these extreme cases
- add checks to `msg_server_liquid_stake.go` and `msg_server_redeem_stake.go`
- added quick unit test cases for these bounds

## Testing and Verifying

1. Manual integration test

To test, set the starting RR to an extreme value and run the integration tests to observe the failure when liquid staking or redeeming. _Note: I had to set the invariant check interval to `0` (which means "don't check the invariants at all except at genesis") in order to test with these extreme value redemption rates (check `app.go:362`)_

Checks:
A) set the RR to 0.1 with `RedemptionRate:     sdk.NewDec(1).Quo(sdk.NewDec(10))`
=> the integration tests should "hang" on liquid staking and you'll see this log:
```
11:01PM ERR Redemption rate 0.100000000000000000 is below min safety threshold 0.900000000000000000 module=x/stakeibc
```
B) set the RR to 10 with `	RedemptionRate:     sdk.NewDec(10)`
=> the integration tests should fail on redeeming!
```
11:26PM ERR Redemption rate 10.000000000000000000 is above max safety threshold 3.000000000000000000 module=x/stakeibc
```

3. Run the unit tests, they have cases for this min and max now!

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not documented
  - Does this pull request update existing proto field values (and require a backend and frontend migration)? no
  - Does this pull request change existing proto field names (and require a frontend migration)? no, **but we add new proto fields!**

### CODE REVIEW

File: x/stakeibc/keeper/msg_server_register_host_zone.go [hunk 1]
```diff
@@ -50,7 +50,7 @@ func (k msgServer) RegisterHostZone(goCtx context.Context, msg *types.MsgRegiste
 		HostDenom:         msg.HostDenom,
 		TransferChannelId: msg.TransferChannelId,
 		// Start exchange rate at 1 upon registration
-		RedemptionRate:     sdk.NewDec(1),
+		RedemptionRate:     sdk.NewDec(10),
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_register_host_zone.go:53]: Should the base (success) case have RedemptionRate == 1?

**Review by asalzmann — CHANGES_REQUESTED** [review]
lgtm overall, thanks for adding this! one question

Is there any reason to split out `IsRedemptionRateAboveMinSafetyThreshold` and `IsRedemptionRateBelowMaxSafetyThreshold`, vs. checking that the rate is between the bounds? It might make it slightly easier to reason about the system overall if we merged the two functions (you'd never want to liquid stake with a high RR / redeem with a low RR, but it could happen with the current setup)

File: x/stakeibc/keeper/msg_server_register_host_zone.go (hunk 1 above)
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_register_host_zone.go:53]: yes, shifted back to 1

File: x/stakeibc/keeper/msg_server_liquid_stake.go [hunk 2]
```diff
@@ -51,6 +51,13 @@ func (k msgServer) LiquidStake(goCtx context.Context, msg *types.MsgLiquidStake)
 		return nil, sdkerrors.Wrapf(types.ErrInvalidToken, "denom is not an IBC token (%s)", ibcDenom)
 	}
 
+	// safety check: redemption rate must be above safety threshold
+	rateIsSafe, error := k.IsRedemptionRateAboveMinSafetyThreshold(ctx, *hostZone)
```
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_liquid_stake.go:55]: done

File: x/stakeibc/keeper/msg_server_redeem_stake.go [hunk 3]
```diff
@@ -57,6 +57,13 @@ func (k msgServer) RedeemStake(goCtx context.Context, msg *types.MsgRedeemStake)
 		return nil, sdkerrors.Wrapf(types.ErrInvalidAmount, fmt.Sprintf("invalid amount: %s", err.Error()))
 	}
 
+	// safety check: redemption rate must be above safety threshold
+	rateIsSafe, error := k.IsRedemptionRateBelowMaxSafetyThreshold(ctx, hostZone)
```
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_redeem_stake.go:61]: done

File: x/stakeibc/types/params.go [hunk 4]
```diff
@@ -19,24 +19,27 @@ var (
...
-	DefaultICATimeoutNanos  uint64 = 600000000000
-	DefaultBufferSize       uint64 = 5   // 1/5=20% of the epoch
-	DefaultIbcTimeoutBlocks uint64 = 300 // 300 blocks ~= 30 minutes
-	DefaultFeeTransferTimeoutNanos  uint64 = 600000000000 // 10 minutes
-
+	DefaultICATimeoutNanos                  uint64 = 600000000000
+	DefaultBufferSize                       uint64 = 5            // 1/5=20% of the epoch
+	DefaultIbcTimeoutBlocks                 uint64 = 300          // 300 blocks ~= 30 minutes
+	DefaultFeeTransferTimeoutNanos          uint64 = 600000000000 // 10 minutes
+	DefaultSafetyMinRedemptionRateThreshold uint64 = 90           // divide by 100, so 90 = 0.9
```
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/types/params.go:26]: agreed, let's put this in the "high risk" bucket

**Comment by riley-stride** [comment]
> lgtm overall, thanks for adding this! one question
> 
> Is there any reason to split out `IsRedemptionRateAboveMinSafetyThreshold` and `IsRedemptionRateBelowMaxSafetyThreshold`, vs. checking that the rate is between the bounds? It might make it slightly easier to reason about the system overall if we merged the two functions (you'd never want to liquid stake with a high RR / redeem with a low RR, but it could happen with the current setup)

No reason. Merged into a single function in latest commit and checked in both liquid_stake and redeem_stake msg servers!

**Comment by asalzmann** [comment]
One last minor point - should we validate that the params are within a range? E.g. `DefaultSafetyMinRedemptionRateThreshold` is g.t. 80 (or some other reasonable number?)

Similar to https://github.com/Stride-Labs/stride/blob/main/x/stakeibc/types/params.go#L127

**Review by asalzmann — APPROVED** [review]

**Comment by riley-stride** [comment]
> One last minor point - should we validate that the params are within a range? E.g. `DefaultSafetyMinRedemptionRateThreshold` is g.t. 80 (or some other reasonable number?)
> 
> Similar to https://github.com/Stride-Labs/stride/blob/main/x/stakeibc/types/params.go#L127

Good idea, done

**Pull request merged**